File: lc/T101-150/t_107.java


package leetcoder;

import java.util.*;
import java.util.concurrent.LinkedBlockingDeque;
import java.util.concurrent.LinkedBlockingQueue;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>> list = new ArrayList<>();
        if(root == null) return list;
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        Stack<List<Integer>> s = new Stack<>();
        while(!q.isEmpty()){
            List<Integer> lis = new ArrayList<>();
            Queue<TreeNode> t = new LinkedList<>();
            while(!q.isEmpty()){
                TreeNode tmp = ((LinkedList<TreeNode>) q).pollFirst();
                lis.add(tmp.val);
                if(tmp.left  !=null)
                    t.add(tmp.left);

                if(tmp.right !=null)
                    t.add(tmp.right);

            }
            s.push(lis);
            q=t;
        }
        while(!s.empty()){
            list.add(s.pop());
        }
        return list ;


    }
}
















class Main {

    public static void main(String[] arg) {

        TreeNode head = new TreeNode(1);
        TreeNode head1 = new TreeNode(2);
        TreeNode head2 = new TreeNode(3);
        head.left = head1;
        head1.right = head2;
        new leetcoder.Solution().levelOrderBottom(head);



    }
}

File: lc/T101-150/T107.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
      vector<vector<int>> res;
      travel(res,root,0);
      return res;
    }
    void travel(    vector<vector<int>> & res ,TreeNode* root ,int level ){
        if(root==nullptr) return;
        if(res[level].size()==0) res[level] = {};
        res[level].emplace_back(root->val);
         travel(res,root->left,level+1);
         travel(res,root->right,level+1);
        return;
    }
};